File: legacy/ui/features_v01_test.go
package ui

import (
	"testing"
	"github.com/zyedidia/rope"
	assert "github.com/stretchr/testify/assert"
)

func TestOnCommentLine(t *testing.T) {
	e.code = rope.New([]byte("this is a sample text"))

	apply_highlighter(e.code.Value(), "", "")

	e.langConf.Comment = "//"
	e.OnCommentLine()

	expected := "//this is a sample text"
	assert.Equal(t, expected, string(e.code.Value()), "comment line mismatch")

	e.OnUndo()
	actual := "this is a sample text"
	assert.Equal(t, actual, string(e.code.Value()), "undo comment line mismatch")

	e.OnRedo()
	assert.Equal(t, expected, string(e.code.Value()), "redo comment line mismatch")
}

func TestOnSwapLinesUp(t *testing.T) {
	e.code = rope.New([]byte("first line of text\nsecond line of text\nthird line of text\n"))

	e.Row = 1
	e.Col = 0

	apply_highlighter(e.code.Value(), "", "")
	e.OnSwapLinesUp()

	expected := "second line of text\nfirst line of text\nthird line of text\n"
	assert.Equal(t, expected, string(e.code.Value()), "swaplinesup mismatch")

	e.OnUndo()
	actual := "first line of text\nsecond line of text\nthird line of text\n"
	assert.Equal(t, actual, string(e.code.Value()), "undo swaplinesup mismatch")

	e.OnRedo()
	assert.Equal(t, expected, string(e.code.Value()), "redo swaplinesup mismatch")
}

func TestOnSwapLinesDown(t *testing.T) {
	e.code = rope.New([]byte("first line of text\nsecond line of text\nthird line of text\n"))

	e.Row = 1
	e.Col = 0

	apply_highlighter(e.code.Value(), "", "")
	e.OnSwapLinesDown()

	expected := "first line of text\nthird line of text\nsecond line of text\n"
	assert.Equal(t, expected, string(e.code.Value()), "swaplinesup mismatch")

	e.OnUndo()
	actual := "first line of text\nsecond line of text\nthird line of text\n"
	assert.Equal(t, actual, string(e.code.Value()), "undo swaplinesup mismatch")

	e.OnRedo()
	assert.Equal(t, expected, string(e.code.Value()), "redo swaplinesup mismatch")
}

File: source/features_v01_test.cpp
/**
    Copyright (C) 2023 - 2026, edgo authors

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, see <https://www.gnu.org/licenses/>.
*/

#include "editor.h"
#include "highlighter.h"
#include <gtest/gtest.h>

TEST(FeaturesV01Tests, TestOnCommentLine) {
	e.code = rope<char>("this is a sample text");

	APPLY_HIGHLIGHTER

	e.langConf.comment = "//";
	OnCommentLine();

	auto expected = "//this is a sample text";
	EXPECT_EQ(expected, e.code_str());

	OnUndo();
	auto actual = "this is a sample text";
	EXPECT_EQ(actual, e.code_str());

	OnRedo();
	EXPECT_EQ(expected, e.code_str());
}

TEST(FeaturesV01Tests, TestOnSwapLinesUp) {
	e.code = rope<char>("first line of text\nsecond line of text\nthird line of text\n");

	e.row = 1;
	e.col = 0;

	APPLY_HIGHLIGHTER
	OnSwapLinesUp();

	auto expected = "second line of text\nfirst line of text\nthird line of text\n";
	EXPECT_EQ(expected, e.code_str());

	OnUndo();
	auto actual = "first line of text\nsecond line of text\nthird line of text\n";
	EXPECT_EQ(actual, e.code_str());

	OnRedo();
	EXPECT_EQ(expected, e.code_str());
}

TEST(FeaturesV01Tests, TestOnSwapLinesDown) {
	e.code = rope<char>("first line of text\nsecond line of text\nthird line of text\n");

	e.row = 1;
	e.col = 0;

	APPLY_HIGHLIGHTER
	OnSwapLinesDown();

	auto expected = "first line of text\nthird line of text\nsecond line of text\n";
	EXPECT_EQ(expected, e.code_str());

	OnUndo();
	auto actual = "first line of text\nsecond line of text\nthird line of text\n";
	EXPECT_EQ(actual, e.code_str());

	OnRedo();
	EXPECT_EQ(expected, e.code_str());
}
